app/assets/javascripts/admin/resources/services/enterprises.js.coffee

```diff
@@ -38,3 +38,23 @@ angular.module("admin.resources").factory 'Enterprises', ($q, EnterpriseResource
 
     resetAttribute: (enterprise, attribute) ->
       enterprise[attribute] = @pristineByID[enterprise.id][attribute]
+
+    removeLogo: (enterprise) ->
```

Comment:
this is copy pasted twice from save, you can remove that with something like this:
callFunctionOnEnterpriseResource( EnterpriseResource.$update, enterprise );
callFunctionOnEnterpriseResource( EnterpriseResource.removeLogo, enterprise );
callFunctionOnEnterpriseResource( EnterpriseResource.removePromoImage, enterprise );

and then something like:

callFunctionOnEnterpriseResource: (function, enterprise) ->
      deferred = $q.defer()
      function.apply( null, ($update({id: enterprise.permalink})
      .then( (data) =>
        @pristineByID[enterprise.id] = angular.copy(enterprise)
        deferred.resolve(data)
      ).catch (response) ->
        deferred.reject(response))
      deferred.promise

Do you want to give it a try?

Reply:
Nice suggestion, @luisramos0. I did something patterned after this for the two methods, and it works great.

Regarding `Enterprises.save`, I didn't refactor this yet because it uses `angular.copy(enterprise)` (as opposed to `angular.copy(data)`) which means that it would cache whatever has been loaded/changed in the JS controller:

```js
save: (enterprise) ->
  deferred = $q.defer()
  enterprise.$update({id: enterprise.permalink})
  .then( (data) =>
    @pristineByID[enterprise.id] = angular.copy(enterprise)
    deferred.resolve(data)
  ).catch (response) ->
    deferred.reject(response)
  deferred.promise
```

I suspect this is because `admin/enterprises#update` uses `Api::Admin::IndexEnterpriseSerializer` not `Api::Admin::EnterpriseSerializer`, so the returned data is incomplete.

I'm not yet comfortable enough with the code, but it seems to me that it is appropriate to have the server-side action use the non-index serializer. Do you see possible issues from doing this, aside from the difference in performance?..

It would also make sure that the object stored in `@pristineByID` is exactly what is persisted in the server, after any server-side validation or changes.

Reply:
The JS code is exactly the same even if then the things going through it are different If you see a risk in changing it, it's fine! Lets leave it as it is. I'd not change server side just because of this. I was just suggesting possible improvements. Also, function names are my own preference/suggestion, feel free to ignore.

Reply:
@luisramos0 The enterprise settings page loads data with `EnterpriseSerializer`.

`Enterprises.save` caches `angular.copy(enterprise)` (UI-updated copy) and disregards the returned value from the API, which seems well because `admin/enterprises#update` uses `IndexEnterpriseSerializer` not `EnterpriseSerializer`.

The `remove_*` endpoints uses `EnterpriseSerializer`, and update the cached enterprise with `angular.copy(data)` (fresh API copy).

Thanks. :slightly_smiling_face: Happy if you're okay with leaving this as is for now.

The function name I renamed as you suggested, because I don't remember ever seeing the format I used being used - I was just nitpicking when I wrote the code.

Reply:
awesome, thanks for the detailed explanation!